docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,281 @@
+   For an example of Docker Swarm secrets, see the
+   [docker-compose secrets stanza](https://docs.docker.com/compose/compose-file/#secrets).
+   Secrets distributed in this manner become part of the RaftDB,
+   and thus it becomes necessary to enable swarm autolock mode,
+   which prevents the Raft database encryption key
+   from being stored plaintext on disk.
+
+   * Kubernetes:
+   Kubernetes is not officially supported by the EdgeX project.
+   Kubernetes also supports the secrets volume approach,
+   though the secrets volume can be mounted anywhere in the container namespace.
+   For an example of Kubernetes secrets volumes, see the
+   [Kubernetes secrets documentation](https://kubernetes.io/docs/concepts/configuration/secret/).
+   Secrets distributed in this manner become part of the `etcd` database,
+   and thus it becomes necessary to specify a
+   [KMS provider for data encryption](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/)
+   to prevent `etcd` from storing plaintext versions of secrets.
+
+   * Docker:
+   Use of a RAM-based temporary file system is desirable to protect secrets,
+   but only Linux platforms support `tmpfs` volumes for secret sharing:
+   generally, `/run` on a Linux system is a `tmpfs` volume.
+   For reasons of portability, `/tmp/edgex/secrets` has been chosen as a
+   cross-platform alternative to `/run`:
+   on the MacOS Docker implementation,
+   `/tmp` is a white-listed shared host directory,
+   while `/run` is not.
+   For production hardening, `/tmp` should be replaced by `/run`
+   where it appears in `docker-compose` files.
+   If the container runtime in use does not support host fileystem bind mounts,
+   then a separate Docker secrets volume should be created for each service.
+
+   * Snaps:
+   Note that due to an existing [bug](https://bugs.launchpad.net/snap-confine/+bug/1620442),
+   `$XDG_RUNTIME_DIR` is set to a snap-specific path
+   (`/run/user/1000/snap.`_$SNAP_NAME_`/`),
+   but the directory is **neither created nor accessible to the snap**.
+   Because it may be necessary to share several secrets outside of the snap
+   --such as secret store tokens--
+   it is desirable to place them in a location that can be mapped to
+   a snap content interface, such as `$SNAP_DATA/secrets`.
+   Although not a `tmpfs`, this is the best available option at this time.
+   The snap bug should be escalated in priority.
+
+## Consequences
+
+As the existing implementation is not fully-compliant with this ADR,
+significant scope will be added to current and future EdgeX releases
+in order to bring the project into compliance.
+
+List of needed improvements:
+
+- PKI private keys
+  * All: Move to using Vault as system of origin for the PKI instead of the standalone `security-secrets-setup` utility.
+  * All: Cache the PKI for Consul and Vault on persistent disk; rotate occasionally.
+  * All: Investigate hardware protection of cached Consul and Vault PKI secret keys.
```

Comment:
Can these not be store in Vault itelf? I thought the master Vault token was all we really needed to secure via HW?

Reply:
> Can these not be store in Vault itelf? I thought the master Vault token was all we really needed to secure via HW?

Sadly, no.

Vault is incapable of unsealing its own PKI.  This is because Vault implements its /sys/{un}seal API over HTTP(S), which means the PKI must be available to serve the REST endpoint prior to the Vault master key being supplied to Vault to unseal it.  This means that not only must the Vault master key be protected (to protect all the secrets), but also the private key for Vault's REST endpoint.

Consul is in the same position so long as Vault is configured to use a Consul storage backend.  (Vault needs consul for its persistent store, but consul ideally would get its PKI private key out of Vault, where it running.)  The proposal is to break this dependency so that only Vault needs to be involved in solving the problems with Vault.

Although Vault presently does not support it, even the cheapest USB HSMs support hardware-protected private keys for TLS usage.  For example, in Apache it is easy to specify an HSM-protected private key for TLS: (taken from [here](https://serverfault.com/questions/988982/apache-not-finding-the-hsm-protected-private-key-over-pkcs11)):
```
  SSLCertificateFile /etc/pki/tls/apachecert.pem
  SSLCertificateKeyFile "pkcs11:type=private?pin-value=121212;token=VendorPKCS11;id=%51"
```

Reply:
OK, thanks for the explanation. Are you going to file a separate issue for this work or just fold it into the existing Vault master token impl (https://github.com/edgexfoundry/edgex-go/issues/1919)? Is there an ADR for the HW protected storage of the Vault master token?

Reply:
> Is there an ADR for the HW protected storage of the Vault master token?

The HW protected storage work is close to 18 months old at this point and predates adoption of ADRs.  However, I do have a threat model that I can upstream to edgex-docs now that edgex-docs is now markdown-based.

> Are you going to file a separate issue for this work?

I am debating this.  There are many  possible implementations, one of which could be to just pull the key out of Vault when the Vault is unsealed.  Under this imlementation, it would be part of the bootstrapping ADR work.  I am hesitant to file an issue for something that is at least a year out in implementation.